WalletWasabi/WabiSabi/Backend/DoSPrevention/Prison.cs

```diff
@@ -1,143 +1,101 @@
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics;
 using System.Linq;
+using System.Threading.Channels;
 using NBitcoin;
 using WalletWasabi.WabiSabi.Backend.Models;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Backend.Rounds.CoinJoinStorage;
 
 namespace WalletWasabi.WabiSabi.Backend.DoSPrevention;
 
-/// <summary>
-/// Malicious UTXOs are sent here.
-/// </summary>
 public class Prison
 {
-	public Prison() : this(Enumerable.Empty<Inmate>())
-	{
-	}
+	private ICoinJoinIdStore CoinJoinIdStore { get; }
+	private DoSConfiguration DoSConfiguration { get; }
 
-	public Prison(IEnumerable<Inmate> inmates)
+	public Prison(DoSConfiguration dosConfiguration, ICoinJoinIdStore coinJoinIdStore, IEnumerable<Offender> inmates, ChannelWriter<Offender> channelWriterWriter)
 	{
-		Inmates = inmates.ToDictionary(x => x.Utxo, x => x);
+		DoSConfiguration = dosConfiguration;
+		CoinJoinIdStore = coinJoinIdStore;
+		Offenders = inmates.ToList();
+		NotificationChannelWriter = channelWriterWriter;
 	}
 
-	private Dictionary<OutPoint, Inmate> Inmates { get; }
-	private object Lock { get; } = new object();
-
-	/// <summary>
-	/// To identify the latest change happened in the prison.
-	/// </summary>
-	public Guid ChangeId { get; private set; } = Guid.NewGuid();
+	private ChannelWriter<Offender> NotificationChannelWriter { get; }
+	private List<Offender> Offenders { get; }
+	private object Lock { get; } = new();
 
-	public (int noted, int banned, int longBanned) CountInmates()
+	private void Punish(Offender offender)
 	{
 		lock (Lock)
 		{
-			return (Inmates.Count(x => x.Value.Punishment == Punishment.Noted), Inmates.Count(x => x.Value.Punishment == Punishment.Banned), Inmates.Count(x => x.Value.Punishment == Punishment.LongBanned));
+			Offenders.Add(offender);
 		}
+		NotificationChannelWriter.TryWrite(offender);
```

Comment:
Yes, it can be. Btw, it doesn't fail to punish it, it failed to persist the punishment what mean it will be forgotten after restart, what sucks but it protects the coordinator anyway.